clients/githubrepo/branches.go

```diff
@@ -504,27 +532,19 @@ func applyRepoRules(branchRef *clients.BranchRef, rules []*repoRuleSet) {
 }
 
 func translatePullRequestRepoRule(base *clients.BranchProtectionRule, rule *repoRule) {
-	if readBoolPtr(rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush) {
-		base.RequiredPullRequestReviews.DismissStaleReviews = rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush
-	}
-	if readBoolPtr(rule.Parameters.PullRequestParameters.RequireCodeOwnerReview) {
-		base.RequiredPullRequestReviews.RequireCodeOwnerReviews = rule.Parameters.PullRequestParameters.RequireCodeOwnerReview
-	}
-	if readBoolPtr(rule.Parameters.PullRequestParameters.RequireLastPushApproval) {
-		base.RequireLastPushApproval = rule.Parameters.PullRequestParameters.RequireLastPushApproval
-	}
-	if reviewerCount := readIntPtr(rule.Parameters.PullRequestParameters.RequiredApprovingReviewCount); reviewerCount > 0 {
-		base.RequiredPullRequestReviews.RequiredApprovingReviewCount = &reviewerCount
-	}
+	base.RequiredPullRequestReviews.DismissStaleReviews = rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush
+	base.RequiredPullRequestReviews.RequireCodeOwnerReviews = rule.Parameters.PullRequestParameters.RequireCodeOwnerReview
+	base.RequireLastPushApproval = rule.Parameters.PullRequestParameters.RequireLastPushApproval
+	base.RequiredPullRequestReviews.RequiredApprovingReviewCount = rule.Parameters.PullRequestParameters.
+		RequiredApprovingReviewCount
```

Comment:
It's related to the issue of having nil values when we actually have the data and they should be true/false. I discussed further on [this comment](https://github.com/ossf/scorecard/pull/3499/files/95b09cbbaf55e8858ff6479334ddb0f4a93b004a#r1391845595).

Reply:
I think this is a bug given the rest of the code.

```go
base.RequiredPullRequestReviews = new(clients.PullRequestReviewRule)
```

Wouldn't this give credit for a PR, when we haven't necessarily seen any PR stuff yet?

Reply:
Whoops didnt realize this was inside translatePullRequestRepoRule. You're right, allocating the struct here is fine. You should still add a nil check before allocation. Otherwise I think this interferes with rule layering. (you could have one rule which enforces 1 reviewer, and one rule which enforces codeowners)

```
if base.RequiredPullRequestReviews == nil {
    base.RequiredPullRequestReviews = new(clients.PullRequestReviewRule)
}
```

additionally, you'll need to revert the other parts, as they also break layering. 
```
	base.RequiredPullRequestReviews.DismissStaleReviews = rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush
	base.RequiredPullRequestReviews.RequireCodeOwnerReviews = rule.Parameters.PullRequestParameters.RequireCodeOwnerReview
	base.RequireLastPushApproval = rule.Parameters.PullRequestParameters.RequireLastPushApproval
	base.RequiredPullRequestReviews.RequiredApprovingReviewCount = rule.Parameters.PullRequestParameters.
		RequiredApprovingReviewCount
```

they just copy over without taking the more strict value. Which is what the older pattern did:
```
if readBoolPtr(rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush) {
		base.RequiredPullRequestReviews.DismissStaleReviews = rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush
	}
```

Reply:
I discussed offline with Spencer and we agreed to keep this section as it currently is in the PR, because of the following reasons:
1. This new code shouldn't interfere with rule layering (i.e., different rule set affecting the same branch), because conflicts between different rules (and also between rules and traditional Branch Protection configs) are managed by the `mergeBranchProtectionRules` function at https://github.com/diogoteles08/scorecard/blob/feat/branch-protection-recognize-rule-changes-only-through-pr/clients/githubrepo/branches.go#L563  . So this new code also contributes to leave the `translatePullRequestRepoRule` with the single responsibility of translating the data, leaving the conflict management to the other function.
2. The reason why I changed the older pattern (last example on Spencer comment https://github.com/ossf/scorecard/pull/3499#discussion_r1416213404) is because in case we had a `rule.Parameters.PullRequestParameters.DismissStaleReviewsOnPush` with a false value, we could end up not propagating this value and having a `nil` value on the evaluation part of the code. This was a bug, as we want to leave the `nil` value only for the cases we actually don't have access to the information.

Despite from this, the concern with rule layering is very valid! Therefore I'm also including a new test to cover the case of multiples rule sets affecting a same branch.